File: quiz/src/main/java/com/example/quiz/service/interfaces/RolePermissionService.java
package com.example.quiz.service.interfaces;

import com.example.quiz.model.entity.Permission;
import com.example.quiz.model.entity.Role;
import com.example.quiz.model.entity.User;

import java.util.List;
import java.util.Set;

public interface RolePermissionService {
    
    // Role management
    Role createRole(String authority, String description);
    Role getRoleByAuthority(String authority);
    List<Role> getAllRoles();
    Role updateRole(Long roleId, String authority, String description);
    void deleteRole(Long roleId);
    
    // Permission management
    Permission createPermission(String name, String description, String resource, String action);
    Permission getPermissionByName(String name);
    List<Permission> getAllPermissions();
    List<Permission> getPermissionsByResource(String resource);
    Permission updatePermission(Long permissionId, String name, String description, String resource, String action);
    void deletePermission(Long permissionId);
    
    // Role-Permission assignment
    void assignPermissionToRole(Long roleId, Long permissionId);
    void removePermissionFromRole(Long roleId, Long permissionId);
    void assignPermissionsToRole(Long roleId, Set<Long> permissionIds);
    void removeAllPermissionsFromRole(Long roleId);
    
    // User-Role assignment
    void assignRoleToUser(Long userId, Long roleId);
    void removeRoleFromUser(Long userId, Long roleId);
    void assignRolesToUser(Long userId, Set<Long> roleIds);
    void removeAllRolesFromUser(Long userId);
    
    // Utility methods
    List<Role> getUserRoles(Long userId);
    List<Permission> getUserPermissions(Long userId);
    List<User> getUsersByRole(String authority);
    boolean userHasRole(Long userId, String authority);
    boolean userHasPermission(Long userId, String resource, String action);
}

File: quiz-be/src/role-permissions/role-permission.service.ts
import { Injectable } from '@nestjs/common';
import { ROLE_PERMISSION_ERRORS } from '../common/business-errors/role-permission-errors';
import { BusinessException } from '../common/exceptions/business.exception';
import { PrismaService } from '../prisma.service';

@Injectable()
export class RolePermissionService {
  constructor(private readonly prisma: PrismaService) {}

  async getAllRoles() {
    return this.prisma.role.findMany({
      include: {
        permissions: {
          include: {
            permission: true,
          },
        },
      },
      orderBy: { id: 'asc' },
    }).then((roles) =>
      roles.map((r) => ({
        ...r,
        permissions: r.permissions.map((p) => p.permission),
      })),
    );
  }

  async createRole(authority: string, description?: string) {
    return this.prisma.role.create({
      data: { authority, description },
    });
  }

  async updateRole(id: number, authority: string, description?: string) {
    return this.prisma.role.update({
      where: { id },
      data: { authority, description },
    });
  }

  async deleteRole(id: number) {
    await this.prisma.role.delete({ where: { id } });
  }

  async getAllPermissions() {
    return this.prisma.permission.findMany({ orderBy: { id: 'asc' } });
  }

  async createPermission(
    name: string,
    description: string,
    resource: string,
    action: string,
  ) {
    return this.prisma.permission.create({
      data: { name, description, resource, action },
    });
  }

  async updatePermission(
    id: number,
    name: string,
    description: string,
    resource: string,
    action: string,
  ) {
    return this.prisma.permission.update({
      where: { id },
      data: { name, description, resource, action },
    });
  }

  async deletePermission(id: number) {
    await this.prisma.permission.delete({ where: { id } });
  }

  async assignPermissionsToRole(roleId: number, permissionIds: number[]) {
    await this.prisma.rolePermission.deleteMany({ where: { role_id: roleId } });

    if (permissionIds.length > 0) {
      await this.prisma.rolePermission.createMany({
        data: permissionIds.map((permissionId) => ({
          role_id: roleId,
          permission_id: permissionId,
        })),
        skipDuplicates: true,
      });
    }

    return this.getRoleWithPermissions(roleId);
  }

  async assignRoleToUser(roleId: number, userId: string) {
    const role = await this.prisma.role.findUnique({ where: { id: roleId } });
    if (!role) {
      throw new BusinessException(ROLE_PERMISSION_ERRORS.ROLE_NOT_FOUND);
    }

    await this.prisma.userRole.create({
      data: { role_id: roleId, user_id: userId },
    });
  }

  async assignUsersToRole(roleId: number, userIds: string[]) {
    await this.prisma.userRole.deleteMany({ where: { role_id: roleId } });
    if (userIds.length > 0) {
      await this.prisma.userRole.createMany({
        data: userIds.map((userId) => ({ user_id: userId, role_id: roleId })),
        skipDuplicates: true,
      });
    }
  }

  async getUsersByRole(authority: string) {
    const role = await this.prisma.role.findUnique({
      where: { authority },
      include: {
        users: {
          include: {
            user: true,
          },
        },
      },
    });

    if (!role) {
      return [];
    }

    return role.users.map((row) => ({
      id: row.user.id,
      username: row.user.username,
      email: row.user.email,
      fullName: row.user.full_name,
    }));
  }

  async userHasPermission(userId: string, resource: string, action: string) {
    const found = await this.prisma.userRole.findFirst({
      where: {
        user_id: userId,
        role: {
          permissions: {
            some: {
              permission: {
                resource,
                action,
              },
            },
          },
        },
      },
    });

    return !!found;
  }

  private async getRoleWithPermissions(roleId: number) {
    const role = await this.prisma.role.findUnique({
      where: { id: roleId },
      include: {
        permissions: {
          include: {
            permission: true,
          },
        },
      },
    });

    if (!role) {
      throw new BusinessException(ROLE_PERMISSION_ERRORS.ROLE_NOT_FOUND);
    }

    return {
      ...role,
      permissions: role.permissions.map((p) => p.permission),
    };
  }
}
